client/client.go

```diff
@@ -163,44 +170,231 @@ func makeURL(base string, parts []string) string {
 	return base + strings.Join(parts, "/")
 }
 
-func (c *client) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *client) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	return makeURL(c.baseURL, parts), nil
 }
 
 func (c *authenticatingClient) SetRequiredServiceTypes(requiredServiceTypes []string) {
 	c.requiredServiceTypes = requiredServiceTypes
 }
 
-func (c *authenticatingClient) SendRequest(method, svcType, apiCall string, requestData *goosehttp.RequestData) (err error) {
-	err = c.sendAuthRequest(method, svcType, apiCall, requestData)
+func (c *authenticatingClient) SendRequest(
+	method, svcType, apiVersion, apiCall string,
+	requestData *goosehttp.RequestData,
+) (err error) {
+	err = c.sendAuthRequest(method, svcType, apiVersion, apiCall, requestData)
 	if gooseerrors.IsUnauthorised(err) {
 		c.setToken("")
-		err = c.sendAuthRequest(method, svcType, apiCall, requestData)
+		err = c.sendAuthRequest(method, svcType, apiVersion, apiCall, requestData)
 	}
 	return
 }
 
-func (c *authenticatingClient) sendAuthRequest(method, svcType, apiCall string, requestData *goosehttp.RequestData) (err error) {
+func (c *authenticatingClient) sendAuthRequest(
+	method, svcType, apiVersion, apiCall string,
+	requestData *goosehttp.RequestData,
+) (err error) {
 	if err = c.Authenticate(); err != nil {
 		return
 	}
 
-	url, err := c.MakeServiceURL(svcType, []string{apiCall})
+	url, err := c.MakeServiceURL(svcType, apiVersion, []string{apiCall})
 	if err != nil {
 		return
 	}
 	return c.sendRequest(method, url, c.Token(), requestData)
 }
 
-func (c *authenticatingClient) MakeServiceURL(serviceType string, parts []string) (string, error) {
+func (c *authenticatingClient) MakeServiceURL(serviceType, apiVersion string, parts []string) (string, error) {
 	if !c.IsAuthenticated() {
 		return "", errors.New("cannot get endpoint URL without being authenticated")
 	}
-	url, ok := c.serviceURLs[serviceType]
+	serviceURL, ok := c.serviceURLs[serviceType]
 	if !ok {
 		return "", errors.New("no endpoints known for service type: " + serviceType)
 	}
-	return makeURL(url, parts), nil
+	if serviceType == "object-store" {
+		// object-store has no versioning capability.
+		return makeURL(serviceURL, parts), nil
+	}
+	if apiVersion == "" {
+		// some tests don't include a version
+		return makeURL(serviceURL, parts), nil
+	}
+	if strings.Contains(serviceURL, c.serviceURLs["object-store"]) {
```

Comment:
No, users can add endpoints with whatever URL they want, with a service type of whatever they want.  One way to pass simplestreams data to Juju is to put it in a Swift container and add an endpoint. Juju explicitly looks for an endpoint with service type "product-streams", and it's tests use other made up names.  We do know that Swift containers are used, so the root url will be the same as the "object-store" url.